added functions to test ANOVA assumptions

mattansb opened the pull request:
These functions test:

- Homogeneity of variance between groups.
- Sphericity for within-subject factors.
- Normality of residuals (via qq-plots) for each error term.

see discussion at #55

singmann commented:
A few comments.

(1) In the current version the code still fails for some situations, where it should now. For example, for me all calls to `residuals_qqplot()` fail with:
```
Error in ggplot(plot_data, aes(sample = x)) : 
  could not find function "ggplot"
```

The following code goes through a few examples and shows where they fail.
```

## purely within
data(md_12.1)
md_1 <- aov_ez("id", "rt", md_12.1, within = "angle")
test_levene(md_1) ## fails
test_sphericity(md_1)
residuals_qqplot(md_1)  ## fails

md_2 <- aov_ez("id", "rt", md_12.1, within = c("angle", "noise"))
test_levene(md_2) ## fails
test_sphericity(md_2)
residuals_qqplot(md_2) ## fails
# Error in ggplot(plot_data, aes(sample = x)) : 
#   could not find function "ggplot"

### purely between
data(obk.long, package = "afex")
between_1 <- aov_car(value ~ treatment + Error(id), data = obk.long)
test_levene(between_1)
test_sphericity(between_1) ## not particularly helpful output
residuals_qqplot(between_1) ## fails
# Error in ggplot(plot_data, aes(sample = x)) : 
#   could not find function "ggplot"

between_2 <- aov_car(value ~ treatment*gender + Error(id), data = obk.long)
test_levene(between_2)
test_sphericity(between_2) ## not particularly helpful output
residuals_qqplot(between_2)
# Error in ggplot(plot_data, aes(sample = x)) : 
#   could not find function "ggplot"

### mixed
mixed_1 <- aov_car(value ~ treatment * gender + Error(id/(phase*hour)), data = obk.long)
test_levene(mixed_1)
test_sphericity(mixed_1)
residuals_qqplot(mixed_1) ## fails
# Error in ggplot(plot_data, aes(sample = x)) : 
#   could not find function "ggplot"
```

If the pull request is accepted, the package should have corresponding tests in the `tests` folder. 

(2) The code is missing examples.

(3) I am really not sure about the residuals plot function. There are multiple reasons for that.
- The package should have `residuals` method for `afex_aov` objects as outlined in #64. This method should be independent of the QQ-plot function. It might be an idea to add this as well, but this is secondary to the `residuals` method.
- I know I told you to find a way of how to deal with the possibility of having no `aov` slot, but my idea was not to recreate it inside the function, but to find a way utilizing the `lm` slot (which is an `mlm` object in case of repeated measures). One problem is that the `aov` calculation can be quite expensive. Another is that I want to remove the reliance on `aov` objects in `afex`.

Maybe a good idea forward is to take the residuals function out of this pull request for now and focus on the other two functions. They need to become more robust (i.e., produce reasonable output for all cases) and need examples and tests. As far as I know you can simply make changes to your `master` branch and they will be reflected in this pull request (i.e., you do not need to create a new one).

mattansb commented:
I've removed the QQ-plot function for now - I'll try to see if I can use the `lm` object for this (and for the residuals?). This would take some work, I think, as the `residuals(afex_object$lm)` for `mlm` object is itself multivariate, and would need to be correctly "sliced" to get the residuals for each within-subject factor (and the between subject residuals).

Regarding the two test functions, I pushed a fix to the error(s). Specifically:

- `test_sphericity` returns an informative error when provided with between only models.
- same for `test_levene` for within only models.

I also added an example usage for each.

singmann commented:
Sorry for being annoying, but the examples should not be in `dontrun` blocks. Instead, they should be actually be able to run (and run automatically when called with `example`).

mattansb commented:
How's that?

singmann commented on R/test_assumption.R:
```diff
@@ -30,76 +44,32 @@ test_levene <- function(afex_aov, center = mean, ...){
-residuals_qqplot <- function(afex_aov, return = 'plot') {
-  if (!requireNamespace("ggplot2", quietly = TRUE)) {
-    stop("package ggplot2 is required.", call. = FALSE)
+  if (length(attr(between_1,'within'))==0) {
```
It should probably not be `between_1` here. That suggests you did not check whether the function actually works before making the commit. 

To ensure functions work, all functionality should be accompanied by corresponding unit tests which can be found in the `tests/testthat` folder. I am preparing them for this functionality.

singmann commented:
As you see, I did a few changes. I feel like now this functionality is ready to be added to `afex`. 

The most important change for me is the addition of unit tests in `tests/testthat`. From my long experience of developing `R` packages I have learned painfully, that I am not accepting functionality that does not include a number of simple checks to make sure it actually works.  The reason is clear in the current pull request which was not working as it stands. So please always test what you commit to avoid unnecessary work from my side. In the way you submitted it,  `test_sphericity` always failed.

I also combined functions into one help page and made the examples self-contained (your examples required the existence of `obk.long` in the global environment, which should not be assumed) and moved them into a different file.

Let me know if you want to add the residuals functionality now as well or make another pull-request for that.

mattansb commented:
Looks great!
I think this should be added to `afex` and we can continue with the residuals / qq plot in a separate PR (where I will also add some tests)
